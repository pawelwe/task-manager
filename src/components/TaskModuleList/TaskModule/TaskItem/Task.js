import React from 'react';
import clockIcon from '../../../../images/clock-icon.svg';
import { calculateExpiration } from '../../../../utils/utils';
import classes from './Task.scss';

const Task = ({
  name,
  completed,
  toggleTask,
  moduleId,
  id,
  priority,
  creationDate,
  expirationPeriod,
  removeTask,
}) => {
  const expiration = calculateExpiration(creationDate, expirationPeriod);
  const getCreationDate = new Date(creationDate);
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  const year = getCreationDate.getFullYear();
  const month = months[getCreationDate.getMonth()];
  const date = getCreationDate.getDate();
  const hour = getCreationDate.getHours();
  const min = getCreationDate.getMinutes();

  return (
    <li
      className={`${classes.TaskItemList_item} fadeIn`}
      style={{ textDecoration: completed ? 'line-through' : '' }}
    >
      <span
        id='removeTask'
        onClick={() => removeTask(moduleId, id)}
        className={`removeBtn removeBtn__small`}
      />
      <section
        id='toggleTask'
        className={`${classes.TaskItemList_item_content} tooltip`}
        onClick={() => toggleTask(moduleId, id)}
        data-title='Click to toggle completion...'
      >
        <h5 className={classes.TaskItemList_item_header}>{name}</h5>
        <p className={classes.TaskItemList_item_info}>
          <strong>Created:</strong>{' '}
          {`${date} ${month} ${year} / ${hour}:${min}`}
        </p>
        <p className={classes.TaskItemList_item_info}>
          <strong>Priority:</strong> {priority}
        </p>
        <p className={classes.TaskItemList_item_info}>
          <span className={expiration < 31 ? 'isExpiring' : ''}>
            {expiration >= 31 && (
              <span>
                <strong>Expires in:</strong> {expiration} minutes
              </span>
            )}
            {expiration < 31 &&
              expiration > 0 && (
                <span>
                  <img
                    className='expirationIcon'
                    src={clockIcon}
                    alt='expiration-icon'
                  />
                  Expires in: {expiration} minutes
                </span>
              )}
            {expiration <= 0 && (
              <span>
                <img
                  className='expirationIcon'
                  src={clockIcon}
                  alt='expiration-icon'
                />
                <strong>Expired!</strong>
              </span>
            )}
          </span>
        </p>
      </section>
    </li>
  );
};

export default Task;
