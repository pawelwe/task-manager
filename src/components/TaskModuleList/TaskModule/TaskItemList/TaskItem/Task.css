.TaskItemList_item {
  position: relative;
  margin-bottom: 1rem;
  padding-bottom: 1rem;
  border-bottom: 1px dashed #c2c2c2;
  cursor: pointer;
  user-select: none; }
  .TaskItemList_item:last-of-type {
    border-bottom: 0; }
  .TaskItemList_item_content {
    max-width: calc(100% - 5rem); }
  .TaskItemList_item_header {
    font-size: 1.4rem;
    font-weight: bold;
    margin-bottom: 0.4rem; }
  .TaskItemList_item_info {
    font-size: 1.1rem;
    margin-bottom: 0.2rem; }
    .TaskItemList_item_info:last-of-type {
      margin-top: 0.7rem; }
