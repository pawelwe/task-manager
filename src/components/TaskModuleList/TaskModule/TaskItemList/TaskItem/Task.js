import React from 'react';
import clockIcon from '../../../../../images/clock-icon.svg';
import {
  calculateExpiration,
  calculateAlert,
  months,
} from '../../../../../utils/utils';
import classes from './Task.scss';

const Task = ({
  name,
  completed,
  toggleTask,
  moduleId,
  id,
  priority,
  creationDate,
  expirationPeriod,
  timeFrame,
  removeTask,
  toggleTaskEditMode,
  editMode,
}) => {
  const expiration = calculateExpiration(
    creationDate,
    expirationPeriod,
    timeFrame,
  ).value;
  const expirationAlertMinutesTimeout = 30;
  const timeToTurnOnAlert = calculateAlert(
    expiration,
    timeFrame,
    expirationAlertMinutesTimeout,
  );

  const getCreationDate = new Date(creationDate);
  const year = getCreationDate.getFullYear();
  const month = months[getCreationDate.getMonth()];
  const date = getCreationDate.getDate();
  const hour = getCreationDate.getHours();
  const min = getCreationDate.getMinutes();

  return (
    <li className={classes.TaskItemList_item}>
      <div className={classes.TaskItemList_item_toolbar}>
        {!editMode && (
          <span
            onClick={() => toggleTaskEditMode(moduleId, id, true)}
            data-title="Edit task"
            className={`${classes.editIcon} tooltip`}
          >
            ✏
          </span>
        )}

        {editMode && (
          <span
            className={`${classes.TaskItemList_item_editMode} tooltip`}
            data-title="Task in edit mode"
          >
            ♺
          </span>
        )}

        {!editMode && (
          <span
            id="removeTask"
            data-title="Remove task"
            onClick={() => {
              toggleTaskEditMode(moduleId, id, false);
              removeTask(moduleId, id);
            }}
            className={`${classes.TaskItemList_item_removeBtn} tooltip`}
          >
            ✗
          </span>
        )}
      </div>
      <section
        id="toggleTask"
        className={`${classes.TaskItemList_item_content} tooltip`}
        onClick={() => toggleTask(moduleId, id)}
        style={{ textDecoration: completed ? 'line-through' : '' }}
        data-title="Click to toggle completion..."
      >
        <h5 className={classes.TaskItemList_item_header}>{name}</h5>
        <p className={classes.TaskItemList_item_info}>
          <strong>Created:</strong>{' '}
          {`${date} ${month} ${year} / ${hour}:${min}`}
        </p>
        <p className={classes.TaskItemList_item_info}>
          <strong>Priority:</strong> {priority}
        </p>
        <p className={classes.TaskItemList_item_info}>
          <span className={timeToTurnOnAlert ? 'isExpiring' : ''}>
            {!timeToTurnOnAlert &&
              expiration > 0 && (
                <span>
                  <strong>Expires in:</strong> {expiration} {timeFrame}
                </span>
              )}
            {timeToTurnOnAlert &&
              expiration > 0 && (
                <span>
                  <img
                    className="expirationIcon"
                    src={clockIcon}
                    alt="expiration-icon"
                  />
                  Expires in: {expiration} {timeFrame}
                </span>
              )}
            {expiration <= 0 && (
              <span>
                <img
                  className="expirationIcon"
                  src={clockIcon}
                  alt="expiration-icon"
                />
                <strong>Expired!</strong>
              </span>
            )}
          </span>
        </p>
      </section>
    </li>
  );
};

export default Task;
