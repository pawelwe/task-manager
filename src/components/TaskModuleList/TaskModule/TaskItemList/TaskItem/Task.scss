.TaskItemList_item {
  display: flex;
  justify-content: space-between;
  position: relative;
  margin-bottom: 1rem;
  padding-bottom: 1rem;
  border-bottom: 1px dashed $medium-gray-4;

  user-select: none;
  &:last-of-type {
    border-bottom: 0;
  }

  &_content {
    //max-width: calc(100% - 5rem);
    flex-basis: 100%;
    cursor: pointer;
  }

  &_header {
    font-size: 1.4rem;
    font-weight: bold;
    margin-bottom: 0.4rem;
  }

  &_info {
    font-size: 1.1rem;
    margin-bottom: 0.2rem;
    &:last-of-type {
      margin-top: 0.7rem;
    }
  }

  &_toolbar {
    display: flex;
    order: 1;
  }

  &_editMode {
    font-size: 1.15rem;
    color: $black;
    cursor: help;
  }

  &_removeBtn {
    font-size: 1.3rem;
    cursor: pointer;
    color: $accent;

    &:hover {
      color: lighten($accent, 25%);
    }
  }
}

.editIcon,
.confirmIcon {
  color: $accent;
  cursor: pointer;
  font-size: 1.4rem;
  margin-right: 0.6rem;
}
