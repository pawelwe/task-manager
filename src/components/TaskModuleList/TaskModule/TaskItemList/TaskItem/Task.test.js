import React from 'react';
import { shallow } from 'enzyme';
import Task from './Task';

describe('Task Component', () => {
  const removeTask = jest.fn();
  const toggleTask = jest.fn();
  const wrapper = shallow(<Task removeTask={removeTask} toggleTask={toggleTask} />);

  it('Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('removeTask called', () => {
    wrapper.find('#removeTask').simulate('click');
    expect(removeTask).toHaveBeenCalled();
  });

  it('toggleTask called', () => {
    wrapper.find('#toggleTask').simulate('click');
    expect(toggleTask).toHaveBeenCalled();
  });

});
