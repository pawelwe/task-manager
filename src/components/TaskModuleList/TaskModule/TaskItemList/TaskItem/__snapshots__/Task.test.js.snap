// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Component Snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Task
    removeTask={[Function]}
    toggleTask={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="removeBtn removeBtn__small"
          id="removeTask"
          onClick={[Function]}
/>,
        <section
          className="undefined tooltip"
          data-title="Click to toggle completion..."
          id="toggleTask"
          onClick={[Function]}
>
          <h5
                    className={undefined}
          />
          <p
                    className={undefined}
          >
                    <strong>
                              Created:
                    </strong>
                     
                    NaN undefined NaN / NaN:NaN
          </p>
          <p
                    className={undefined}
          >
                    <strong>
                              Priority:
                    </strong>
                     
          </p>
          <p
                    className={undefined}
          >
                    <span
                              className=""
                    >
                              
                    </span>
          </p>
</section>,
      ],
      "className": "undefined fadeIn",
      "style": Object {
        "textDecoration": "",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "removeBtn removeBtn__small",
          "id": "removeTask",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h5
              className={undefined}
/>,
            <p
              className={undefined}
>
              <strong>
                            Created:
              </strong>
               
              NaN undefined NaN / NaN:NaN
</p>,
            <p
              className={undefined}
>
              <strong>
                            Priority:
              </strong>
               
</p>,
            <p
              className={undefined}
>
              <span
                            className=""
              >
                            
              </span>
</p>,
          ],
          "className": "undefined tooltip",
          "data-title": "Click to toggle completion...",
          "id": "toggleTask",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "h5",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                  Created:
</strong>,
                " ",
                "NaN undefined NaN / NaN:NaN",
              ],
              "className": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Created:",
                },
                "ref": null,
                "rendered": "Created:",
                "type": "strong",
              },
              " ",
              "NaN undefined NaN / NaN:NaN",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                  Priority:
</strong>,
                " ",
                undefined,
              ],
              "className": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Priority:",
                },
                "ref": null,
                "rendered": "Priority:",
                "type": "strong",
              },
              " ",
              undefined,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span
                className=""
>
                
</span>,
              "className": undefined,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  false,
                  false,
                  false,
                ],
                "className": "",
              },
              "ref": null,
              "rendered": Array [
                false,
                false,
                false,
              ],
              "type": "span",
            },
            "type": "p",
          },
        ],
        "type": "section",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="removeBtn removeBtn__small"
            id="removeTask"
            onClick={[Function]}
/>,
          <section
            className="undefined tooltip"
            data-title="Click to toggle completion..."
            id="toggleTask"
            onClick={[Function]}
>
            <h5
                        className={undefined}
            />
            <p
                        className={undefined}
            >
                        <strong>
                                    Created:
                        </strong>
                         
                        NaN undefined NaN / NaN:NaN
            </p>
            <p
                        className={undefined}
            >
                        <strong>
                                    Priority:
                        </strong>
                         
            </p>
            <p
                        className={undefined}
            >
                        <span
                                    className=""
                        >
                                    
                        </span>
            </p>
</section>,
        ],
        "className": "undefined fadeIn",
        "style": Object {
          "textDecoration": "",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "removeBtn removeBtn__small",
            "id": "removeTask",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h5
                className={undefined}
/>,
              <p
                className={undefined}
>
                <strong>
                                Created:
                </strong>
                 
                NaN undefined NaN / NaN:NaN
</p>,
              <p
                className={undefined}
>
                <strong>
                                Priority:
                </strong>
                 
</p>,
              <p
                className={undefined}
>
                <span
                                className=""
                >
                                
                </span>
</p>,
            ],
            "className": "undefined tooltip",
            "data-title": "Click to toggle completion...",
            "id": "toggleTask",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "h5",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <strong>
                    Created:
</strong>,
                  " ",
                  "NaN undefined NaN / NaN:NaN",
                ],
                "className": undefined,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Created:",
                  },
                  "ref": null,
                  "rendered": "Created:",
                  "type": "strong",
                },
                " ",
                "NaN undefined NaN / NaN:NaN",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <strong>
                    Priority:
</strong>,
                  " ",
                  undefined,
                ],
                "className": undefined,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Priority:",
                  },
                  "ref": null,
                  "rendered": "Priority:",
                  "type": "strong",
                },
                " ",
                undefined,
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span
                  className=""
>
                  
</span>,
                "className": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    false,
                    false,
                    false,
                  ],
                  "className": "",
                },
                "ref": null,
                "rendered": Array [
                  false,
                  false,
                  false,
                ],
                "type": "span",
              },
              "type": "p",
            },
          ],
          "type": "section",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
